### PR TITLE
[MRG] Gid offset

### LINKED ISSUE #454: add_drives_from_params  vs add_evoked_drive
Hi,

I am trying to use the same parameters to run a simulation with two different methods but not getting the same result.  

The first method is where I add the drives from params:
```python
net = hnn_core.calcium_model(params, add_drives_from_params = True)
```
The second method I add the drives manually (using same parameters from the param file:
```python
net = hnn_core.calcium_model(params, add_drives_from_params = False)
weights_ampa_p1 = {'L2_basket': 0.997291, 'L2_pyramidal':0.990722,'L5_basket':0.614932, 'L5_pyramidal': 0.004153}
weights_nmda_p1 = {'L2_basket': 0.984337, 'L2_pyramidal':1.714247,'L5_basket':0.061868, 'L5_pyramidal': 0.010042}
synaptic_delays_prox = {'L2_basket': 0.1, 'L2_pyramidal': 0.1,'L5_basket': 1., 'L5_pyramidal': 1.}
net1.add_evoked_drive('evprox1', mu=54.897936, sigma=5.401034, numspikes=1, weights_ampa=weights_ampa_p1, weights_nmda=weights_nmda_p1, location='proximal', event_seed = -14, synaptic_delays=synaptic_delays_prox)
```   

I am not getting the same results from both, the dipole is different between both. I tracked the add_drives_from_params function and I found that it substract 18 from the event seed so I adjusted that. The rest appears to be the same, not sure where the problem

### PR DESCRIPTION
**Pull request by jasmainak**
closes #454 

Examples: https://2122-168215891-gh.circle-artifacts.com/0/html/index.html

This PR makes the simulation output agnostic of the order in which the drives are added. The seeds of the drives have to be adjusted for the new scheme.

Another problem that is fixed in this new scheme is that negative seeds are no longer allowed (as in numpy and other software). Earlier a negative seed would be okay because the final seed used would be `event_seed + gid` and `gid` may start from a large positive number instead of 0

### CODE REVIEW

**Comment by jasmainak** [comment]
does any of you remember how the seed for poisson process in gamma example was determined to be 1079? It's an unusually high number.

File: hnn_core/network.py [hunk 1]
```diff
@@ -965,6 +965,8 @@ def _instantiate_drives(self, tstop, n_trials=1):
             for drive in self.external_drives.values():
                 event_times = list()  # new list for each trial and drive
                 for drive_cell_gid in self.gid_ranges[drive['name']]:
+                    drive_cell_gid_offset = (drive_cell_gid -
```
Comment (ntolley) [inline hnn_core/network.py:968]: Does this mean that all drives default to a seed of `0` when `event_seed` isn't defined?

I just want to make sure we aren't running into another form of #414, where different drives have the same spike times if it's not explicitly set.

Comment (ntolley) [reply hnn_core/network.py:968]: Thinking further, the test for the simulation comparison uses `add_drives_from_params=True`. Two tests come to mind:

1) simulation comparison `add_drives_from_params=False` (perhaps a future PR)
2) Confirmation that adding drives in different orders instantiates the same drive objects, since that was the problem in #454

Comment (jasmainak) [reply hnn_core/network.py:968]: well, `event_seed` is by default 2. This PR just makes sure that giving `event_seed=2` always produces the same result. Now, we could change the default `event_seed` to be `None` and this would be compatible with other software like MNE and sklearn. The behavior when it is `None` would not be reproducible though ...

Comment (jasmainak) [reply hnn_core/network.py:968]: I agree about adding the tests, was just being lazy ;-) Thanks for catching that!

Comment (jasmainak) [reply hnn_core/network.py:968]: also using `None` as default would highlight the fact that multiple trials are needed to make a claim.

**Comment by jasmainak** [comment]
okay test added. MPI tests are spuriously failing sometimes, not sure why. I suggest restarting the CI for now. We'll have to dig in deeper at some point ...

File: hnn_core/tests/test_drives.py [hunk 2]
```diff
@@ -402,3 +403,23 @@ def test_add_drives():
             weights_ampa={'L2_pyramidal': 1.},
             synaptic_delays={'L2_pyramidal': 1.},
             probability={'L2_pyramidal': 2.0})
+
+
+def test_drive_random_state():
```
Comment (ntolley) [inline hnn_core/tests/test_drives.py:408]: This is perfect, super simple and effective

File: hnn_core/network.py (hunk 1 above)
Comment (ntolley) [reply hnn_core/network.py:968]: I agree that emphasizing multiple trials should be more visible. I think we're guilty of not pushing it because it's not useful for debugging purposes.

The more I think about it using `None` as a default does sound appealing. I've definitely been guilty of reading too much into the randomness of a single simulation, and see the result disappear with new seeds.

**Comment by ntolley** [comment]
It's unfortunate this PR requires modifying all the examples, but it's definitely informative for me! "Simulate alpha and beta" was altered:
image

I'll look through the rest to see if there's any others

**Comment by jasmainak** [comment]
Humm interesting. Could the older simulation be considered buggy in this case? See the following:

https://github.com/jonescompneurolab/hnn-core/blob/a761e248ddf360710dd60638269f70361f5d6cb3/examples/workflows/plot_simulate_beta.py#L153-L160

when you do `add_erp_drives` on `net_beta` it is not the same as doing `add_erp_drives` on `net_erp` because the gids of the cells are different, therefore you get different seeds for spikes of the evoked drives in the two cases.

**Comment by jasmainak** [comment]
Oh I see you were talking about the other example. This particular example is not affected because of `sigma=0.` ... touché

File: hnn_core/network.py (hunk 1 above)
Comment (rythorpe) [reply hnn_core/network.py:968]: +1 for this! We might want to consider adding a big fat warning sign on our API doc page that highlights this if we implement it in the next release though.

**Comment by rythorpe** [comment]
> does any of you remember how the seed for poisson process in gamma example was determined to be 1079? It's an unusually high number.

In case this was still a question, I think this number was due to the fact that the poisson drive gids were added after the initial 270 real cell gids were assigned, plus the 3 x 270 gids for each of the three tactile evoked response drives (for a grand total of 1080 gids).

**Review by rythorpe — APPROVED** [review]
Looks good @jasmainak!

**Comment by rythorpe** [comment]
Is there still work to be done adjusting the seed values for the examples or are we good to go @jasmainak @ntolley ?

**Comment by jasmainak** [comment]
I think this PR is good to go unless you spot any discrepancies in the documentation pages

**Comment by ntolley** [comment]
Unfortunately the [LFP](https://jonescompneurolab.github.io/hnn-core/stable/auto_examples/howto/plot_record_extracellular_potentials.html#sphx-glr-auto-examples-howto-plot-record-extracellular-potentials-py) example is slightly different. I suspect this is due to the seeds in `add_erp_drives_to_jones_model()`?
